README.md
# Prometheus Juju Exporter

prometheus-juju-exporter snap collects machines' running status in all models under a Juju controller. It also provides an interface to expose the metrics to Prometheus for storage and further usage.

## Deployment
To get the latest stable version of the snap from Snapstore, run:
```bash
sudo snap install prometheus-juju-exporter
```
To get the latest development version of the snap, build from the source code and install with `--dangerous` flag:
```bash
make build
sudo snap install --dangerous prometheus-juju-exporter.snap
```


docs/changelog.rst
**Changelog**

1.0
^^^
Dec 2022

* Add function to connect to a juju controller to collect machine status data
* Add initial implementation of hardware detection
* Add function to log collected metrics in prometheus registry and expose it as an http server
* Add lint, unit, and functional tests
* Setup GitHub workflow


tests/unit/conftest.py
#!/usr/bin/python3
"""Pytest fixture definitions."""

import os
from unittest import mock

import pytest
from juju.controller import Controller
from juju.model import Model

os.environ["PROMETHEUSJUJUEXPORTERDIR"] = os.path.dirname(os.path.abspath(__file__))


def get_model_list_data():
    """Mock controller.model_uuids."""
    mock_model_list = mock.AsyncMock()

    # uuids are randomly generated for testing purposes
    mock_model_list.return_value = {
        "controller": "65f76aed-789f-4dbf-a75a-a32e5d90ab7e",
        "default": "77643b91-a6f8-4cf6-8755-83c6becd09bb",
        "test": "68c4cc81-9b3d-44a2-a419-d25dfb9d5588",
    }

    return mock_model_list


def get_juju_stats_data():
    """Mock model.get_status."""
    mock_stats = mock.AsyncMock()

    mock_stats.return_value = {
        "model": {
            "name": "test",
            "type": "iaas",
            "controller": "test-cloud-local",
            "cloud": "test-cloud",
            "region": "local",
            "version": "2.9.29",
            "model-status": {"current": "available", "since": "24 Nov 2022 13:19:09Z"},
            "sla": "unsupported",
        },
        "machines": {
            "0": {
                "agent-status": {
                    "status": "started",
                    "since": "24 Nov 2022 13:21:25Z",
                    "version": "2.9.29",
                },
                "hostname": "juju-000ddd-test-0",
                "dns-name": "10.5.0.18",
                "ip-addresses": ["10.5.0.01", "252.0.0.1"],
                "instance-id": "149e81c8-a05b-4852-9201-434670598c30",
                "instance-status": {
                    "status": "running",
                    "message": "ACTIVE",
                    "since": "24 Nov 2022 13:20:33Z",
                },
                "modification-status": {
                    "current": "idle",
                    "since": "24 Nov 2022 13:19:43Z",
                },
                "series": "focal",
                "network-interfaces": {
                    "ens3": {
                        "ip-addresses": ["10.5.0.01"],
                        "mac-address": "fa:16:3e:d4:00:00",
                        "gateway": "10.5.0.1",
                        "space": "alpha",
                        "is-up": "true",
                    },
                    "fan-252": {
                        "ip-addresses": ["252.0.0.1"],
                        "mac-address": "9e:fc:ca:87:00:00",
                        "space": "alpha",
                        "is-up": "true",
                    },
                    "lxdbr0": {
                        "ip-addresses": ["10.100.00.1"],
                        "mac-address": "00:16:3e:e1:00:00",
                        "is-up": "true",
                    },
                },
                "containers": {
                    "0/lxd/0": {
                        "agent-status": {
                            "status": "started",
                            "since": "24 Nov 2022 13:23:50Z",
                            "version": "2.9.29",
                        },
                        "hostname": "juju-000ddd-0-lxd-0",
                        "dns-name": "252.0.0.190",
                        "ip-addresses": ["252.0.0.190"],
                        "instance-id": "juju-000ddd-0-lxd-0",
                        "instance-status": {
                            "status": "running",
                            "message": "Container started",
                            "since": "24 Nov 2022 13:22:46Z",
                        },
                        "modification-status": {
                            "current": "applied",
                            "since": "24 Nov 2022 13:22:46Z",
                        },
                        "series": "focal",
                        "network-interfaces": {
                            "eth0": {
                                "ip-addresses": ["252.0.0.190"],
                                "mac-address": "00:16:3e:7f:00:00",
                                "gateway": "252.0.0.1",
                                "space": "alpha",
                                "is-up": "true",
                            }
                        },
                        "constraints": "arch=amd64 spaces=",
                        "hardware": "availability-zone=nova",
                    }
                },
                "hardware": "arch=amd64 cores=1 mem=2048M root-disk=20480M availability-zone=nova",
            }
        },
        "applications": {
            "ubuntu": {
                "charm": "ubuntu",
                "series": "focal",
                "os": "ubuntu",
                "charm-origin": "charmhub",
                "charm-name": "ubuntu",
                "charm-rev": 21,
                "charm-channel": "stable",
                "exposed": "false",
                "application-status": {
                    "current": "active",
                    "since": "24 Nov 2022 13:23:52Z",
                },
                "units": {
                    "ubuntu/0": {
                        "workload-status": {
                            "current": "active",
                            "since": "24 Nov 2022 13:23:52Z",
                        },
                        "juju-status": {
                            "current": "idle",
                            "since": "24 Nov 2022 13:23:56Z",
                            "version": "2.9.29",
                        },
                        "leader": "true",
                        "machine": "0/lxd/0",
                        "public-address": "252.0.18.190",
                    }
                },
                "version": "20.04",
            }
        },
        "storage": {},
        "controller": {"timestamp": "13:38:17Z"},
    }

    return mock_stats


def collected_stats_data():
    return {
        "example_gauge": {
            "gauge_desc": "This is an example gauge",
            "labels": [
                "job",
                "hostname",
                "customer",
                "cloud_name",
                "juju_model",
                "type",
            ],
            "labelvalues_update": [
                (
                    {
                        "job": "prometheus-juju-exporter",
                        "hostname": "hostname1",
                        "customer": "customer1",
                        "cloud_name": "cloud name",
                        "juju_model": "juju model",
                        "type": "machine type",
                    },
                    0,
                ),
                (
                    {
                        "job": "prometheus-juju-exporter",
                        "hostname": "hostname2",
                        "customer": "customer2",
                        "cloud_name": "cloud name",
                        "juju_model": "juju model",
                        "type": "machine type",
                    },
                    0,
                ),
            ],
            "labelvalues_remove": [
                [
                    "prometheus-juju-exporter",
                    "hostname3",
                    "customer3",
                    "cloud name",
                    "juju model",
                    "machine type",
                ]
            ],
        }
    }


@pytest.fixture
def config_instance(monkeypatch):
    """Mock config."""
    from prometheus_juju_exporter.config import Config

    def _config(args=""):
        # Clear global
        monkeypatch.setattr("prometheus_juju_exporter.config.CONFIG", None)

        mock_set_args = mock.MagicMock()
        monkeypatch.setattr(
            "prometheus_juju_exporter.config.confuse.Configuration.set_args",
            mock_set_args,
        )

        return Config(args)

    return _config


@pytest.fixture
def mock_controller_connection(monkeypatch, mock_model_connection):
    """Mock juju controller for the collector module path."""
    mock_controller = Controller
    mock_connection = mock.AsyncMock()
    mock_connection.return_value = mock_connection
    mock_controller.connect = mock_connection
    mock_controller.disconnect = mock_connection
    mock_controller.is_connected.return_value = False
    mock_controller.model_uuids = get_model_list_data()
    mock_controller.get_model = mock.AsyncMock()
    mock_controller.get_model.return_value = mock_model_connection()
    monkeypatch.setattr(
        "prometheus_juju_exporter.collector.Controller",
        mock_controller,
    )

    return mock_controller


@pytest.fixture
def mock_model_connection():
    """Mock juju model for the collector module path."""
    mock_model = Model
    mock_model.get_status = get_juju_stats_data()

    return mock_model


@pytest.fixture
def exporter_daemon(monkeypatch, stats_gauge):
    """Mock exporter daemon."""
    from prometheus_juju_exporter.exporter import ExporterDaemon

    def _daemon(*args, **kwargs):
        # Clear global
        mock_http_server = mock.MagicMock()
        mock_collector = mock.AsyncMock()
        mock_collector.return_value = collected_stats_data()
        mock_async_sleep = mock.AsyncMock()
        mock_logger = mock.MagicMock()
        mock_logger.return_value = mock_logger
        mock_logger.handlers = None

        monkeypatch.setattr("prometheus_juju_exporter.config.CONFIG", None)
        monkeypatch.setattr(
            "prometheus_juju_exporter.exporter.start_http_server", mock_http_server
        )
        monkeypatch.setattr(
            "prometheus_juju_exporter.collector.Collector.get_stats",
            mock_collector,
        )
        monkeypatch.setattr("prometheus_juju_exporter.logging.getLogger", mock_logger)
        monkeypatch.setattr("asyncio.sleep", mock_async_sleep)
        return ExporterDaemon(*args, **kwargs)

    return _daemon


@pytest.fixture
def collector_daemon(monkeypatch, mock_model_connection, mock_controller_connection):
    """Mock collector."""
    from prometheus_juju_exporter.collector import Collector

    def _collector(*args, **kwargs):
        # Clear global
        monkeypatch.setattr("prometheus_juju_exporter.config.CONFIG", None)
        mock_logger = mock.MagicMock()
        mock_logger.return_value = mock_logger
        mock_logger.handlers = None
        monkeypatch.setattr("prometheus_juju_exporter.logging.getLogger", mock_logger)
        return Collector(*args, **kwargs)

    return _collector


@pytest.fixture
def mock_gauge():
    """Provide mechanism to test args passed in calls to prometheus_client.Gauge."""

    class TestArgs(object):
        def __init__(self):
            self.args = []
            self.kwargs = []
            self.values = []
            self.call_labels = []

        def __call__(self, *args, **kwargs):
            self.args.append(list(args))
            if "registry" in kwargs:
                kwargs.pop("registry")
            self.kwargs.append(dict(kwargs))
            return self

        def labels(self, *args, **kwargs):
            self.call_labels.append(kwargs)
            return self

        def set(self, value):
            self.values.append(value)

        def remove(self, *labelvalues):
            self.values = []

    return TestArgs()


@pytest.fixture
def stats_gauge(monkeypatch, mock_gauge):
    """Mock Gauge calls for the collector module path."""
    monkeypatch.setattr("prometheus_juju_exporter.exporter.Gauge", mock_gauge)
    gauge = mock_gauge

    return gauge


tests/unit/test_cli.py
#!/usr/bin/python3
"""Test Cli."""
import unittest.mock as mock

import pytest

from prometheus_juju_exporter.cli import config_logger, main


class TestCli:
    """Cli test class."""

    @pytest.mark.parametrize("option", ("-d", "--debug"))
    def test_cli_main(self, monkeypatch, option):
        """Test main function in cli."""
        mock_exporter = mock.MagicMock()
        monkeypatch.setattr(
            "prometheus_juju_exporter.cli.ExporterDaemon", mock_exporter
        )

        with mock.patch("prometheus_juju_exporter.cli.config_logger") as config_logger:
            main([option])
            config_logger.assert_called_once_with(debug=True)
            mock_exporter.assert_called_once()

    @pytest.mark.parametrize(
        "arg_option, level_option", [(True, "DEBUG"), (False, "INFO")]
    )
    def test_config_logger(self, arg_option, level_option):
        """Test main function in cli."""
        with mock.patch(
            "prometheus_juju_exporter.cli.project_logger.setLevel"
        ) as set_level:
            config_logger(debug=arg_option)
            set_level.assert_called_once_with(level_option)


tests/unit/test_config.py
#!/usr/bin/python3
"""Test Config class."""
import unittest.mock as mock

import pytest


class TestConfig:
    """Config test class."""

    def test_fixture(self, config_instance):
        """See if the helper fixture works to load charm configs."""
        config_ins = config_instance()
        assert config_ins is not None

    def test_arg(self, config_instance):
        """Check if set_args() is successfully called when arg is not None."""
        config_ins = config_instance({"foo.bar": "car"})
        config_ins.config.set_args.assert_called_once_with(
            {"foo.bar": "car"}, dots=True
        )

    def test_get_config(self, config_instance):
        """Check if getting section config behaves correctly."""
        config_ins = config_instance()
        config_data = config_ins.get_config("juju")
        assert config_data["controller_endpoint"].get() == "192.168.1.100:17070"
        assert (
            config_data["controller_cacert"].get()
            == "-----BEGIN CERTIFICATE-----\n-----END CERTIFICATE-----\n"
        )
        assert config_data["username"].get() == "example_user"
        assert config_data["password"].get() == "example_password"

    def test_validate_config_options_success(self, config_instance):
        """Test validate_config_values function."""
        config_ins = config_instance()

        # Test valid config values in config.yaml file
        with mock.patch("prometheus_juju_exporter.config.exit") as exit_call:
            config_ins.validate_config_options()
            exit_call.assert_not_called()

    @pytest.mark.parametrize("port_value", ("foo", 65536, None))
    def test_validate_config_options_fail(self, config_instance, port_value):
        """Test validate_config_values function."""
        config_ins = config_instance()

        # Test bad config values
        config_ins.config["exporter"]["port"].set(port_value)
        with mock.patch("prometheus_juju_exporter.config.exit") as exit_call:
            config_ins.validate_config_options()
            assert config_ins.config["exporter"]["port"].get() == port_value
            exit_call.assert_called_once()


prometheus_juju_exporter/cli.py
"""CLI module."""
import argparse
from typing import Any

from prometheus_juju_exporter import logger as project_logger
from prometheus_juju_exporter.exporter import ExporterDaemon


def config_logger(debug: bool = False) -> None:
    """Configure global logger's logging level.

    :param bool debug: Whether to set logging level to debug
    """
    project_logger.setLevel("DEBUG" if debug else "INFO")


def main(args: Any = None) -> None:
    """Program entry point.

    Parse cli arguments and start exporter daemon.
    """
    cli = argparse.ArgumentParser(
        prog="prometheus-juju-exporter",
        description="PrometheusJujuExporter CLI",
    )

    cli.add_argument(
        "-d", "--debug", dest="debug", action="store_true", help="Print debug log"
    )

    parser, _ = cli.parse_known_args(args)
    config_logger(debug=parser.debug)

    obj = ExporterDaemon()
    obj.run()


if __name__ == "__main__":  # pragma: no cover
    main()


prometheus_juju_exporter/config.py
"""Configuration loader."""
import sys
from collections import OrderedDict
from logging import getLogger
from typing import Any, Dict, Union

import confuse


class ConfigMeta(type):
    """Singleton metaclass for the Config."""

    _instance = None

    def __new__(cls, *args: Any, **kwargs: Any) -> type:
        if not cls._instance:
            cls._instance = super(ConfigMeta, cls).__new__(cls, *args, **kwargs)
            cls._instance.config = confuse.Configuration(
                "PrometheusJujuExporter", __name__
            )
        return cls._instance


class Config(metaclass=ConfigMeta):
    """Configuration class for PrometheusJujuExporter."""

    config: confuse.Configuration = None

    def __init__(self, args: Union[Dict, None] = None) -> None:
        """Initialize the config class."""
        if not self.config:
            print("Once")
            self.config = confuse.Configuration("PrometheusJujuExporter", __name__)

        self.logger = getLogger(__name__)

        if args:
            self.config.set_args(args, dots=True)

        self.validate_config_options()

    def validate_config_options(self) -> None:
        """Validate the configuration values against a template."""
        template = {
            "exporter": OrderedDict(
                [
                    ("port", confuse.Choice(range(0, 65536), default=5000)),
                    ("collect_interval", int),
                ]
            ),
            "juju": OrderedDict(
                [
                    ("controller_endpoint", str),
                    ("controller_cacert", str),
                    ("username", str),
                    ("password", str),
                ]
            ),
            "customer": OrderedDict([("name", str), ("cloud_name", str)]),
        }

        try:
            self.config.get(template)
            self.logger.info("Configuration parsed successfully")
        except (
            KeyError,
            confuse.ConfigTypeError,
            confuse.ConfigValueError,
            confuse.NotFoundError,
        ) as err:
            self.logger.error("Error parsing configuration values: %s", err)
            sys.exit(1)

    def get_config(self, section: Union[Dict, None] = None) -> confuse.Configuration:
        """Return the config."""
        if section:
            return self.config[section]

        return self.config


prometheus_juju_exporter/exporter.py
"""Exporter module."""
import asyncio
import sys
from logging import getLogger
from typing import Any, Dict, List

from prometheus_client import CollectorRegistry, Gauge, start_http_server

from prometheus_juju_exporter.collector import Collector
from prometheus_juju_exporter.config import Config


class ExporterDaemon:
    """Core class of the exporter daemon."""

    def __init__(self) -> None:
        """Create new daemon and configure runtime environment."""
        self.config = Config().get_config()
        self.logger = getLogger(__name__)
        self.logger.info("Parsed config: %s", self.config.config_dir())
        self._registry = CollectorRegistry()
        self.metrics: Dict[str, Gauge] = {}
        self.collector = Collector()
        self.logger.debug("Exporter initialized")

    def _create_metrics_dict(
        self, gauge_name: str, gauge_desc: str, labels: List[str]
    ) -> None:
        """Create a dict of gauge instances.

        :param str gauge_name: the name of the gauge
        :param str gauge_desc: the description of the gauge
        :param List[str] labels: the label set of the gauge
        """
        if gauge_name not in self.metrics:
            self.logger.debug("Creating Gauge %s", gauge_name)
            self.metrics[gauge_name] = Gauge(
                gauge_name, gauge_desc, labelnames=labels, registry=self._registry
            )

    def update_registry(self, data: Dict[str, Any]) -> None:
        """Update the registry with newly collected values.

        :param dict data: the machine data collected by the Collector method
        """
        for gauge_name, values in data.items():
            self._create_metrics_dict(
                gauge_name=gauge_name,
                gauge_desc=values["gauge_desc"],
                labels=values["labels"],
            )
            for labels, value in values["labelvalues_update"]:
                self.logger.debug(
                    "Updating Gauge %s, %s: %s", gauge_name, labels, value
                )
                self.metrics[gauge_name].labels(**labels).set(value)

            for labels in values["labelvalues_remove"]:
                self.logger.debug(
                    "Deleting labelvalues %s from %s...", labels, gauge_name
                )
                self.metrics[gauge_name].remove(*labels)

    async def trigger(self, **kwargs: Any) -> None:
        """Call Collector and configure prometheus_client gauges from generated stats.

        Available parameters are:

        :param bool once: Whether to continuously collect data. If set
            to true, trigger function will only run once.
        """
        run_collector = True

        while run_collector:
            try:
                self.logger.info("Collecting gauges...")
                data = await self.collector.get_stats()
                self.update_registry(data)
                self.logger.info("Gauges collected and ready for exporting.")
                await asyncio.sleep(
                    self.config["exporter"]["collect_interval"].get(int) * 60
                )
            except Exception as err:  # pylint: disable=W0703
                self.logger.error("Collection job resulted in error: %s", err)
                sys.exit(1)

            # run collector only once if in test mode
            if kwargs.get("once"):
                run_collector = False

    def run(self, **kwargs: Any) -> None:
        """Run exporter.

        Available parameters are:

        :param bool once: Whether to run exporter daemon continuously. If set
            to true, trigger function will only run once.
        """
        self.logger.debug("Running prometheus client http server.")
        start_http_server(
            self.config["exporter"]["port"].get(int),
            registry=self._registry,
        )

        loop = asyncio.get_event_loop()
        task = asyncio.ensure_future(self.trigger(**kwargs))
        loop.run_until_complete(asyncio.wait([task]))


prometheus_juju_exporter/collector.py
"""Collector module."""
from enum import Enum
from logging import getLogger
from typing import Any, Dict, List

from juju.controller import Controller

from prometheus_juju_exporter.config import Config


class MachineType(Enum):
    """String type enum for selecting available machine types."""

    METAL = "metal"
    KVM = "kvm"
    LXD = "lxd"


class Collector:
    """Core class of the PrometheusJujuExporter collector."""

    def __init__(self) -> None:
        """Create new collector and configure runtime environment."""
        self.config = Config().get_config()
        self.logger = getLogger(__name__)
        self.controller = Controller(max_frame_size=6**24)
        self.data: Dict[str, Any] = {}
        self.currently_cached_labels: Dict[str, Any] = {}
        self.previously_cached_labels: Dict[str, Any] = {}
        self.logger.debug("Collector initialized")
        self.prefixes = [
            "52:54:00",
            "fa:16:3e",
            "06:f1:3a",
            "00:0d:3a",
            "00:50:56",
            "fa:16:3e",
        ]

    def refresh_cache(
        self, gauge_name: str, gauge_desc: str, labels: List[str]
    ) -> None:
        """Refresh instances for each collection job.

        :param str gauge_name: the name of the gauge
        :param str gauge_desc: the description of the gauge
        :param List[str] labels: the label set of the gauge
        """
        self.data = {
            gauge_name: {
                "gauge_desc": gauge_desc,
                "labels": labels,
                "labelvalues_update": [],
                "labelvalues_remove": [],
            }
        }

        self.previously_cached_labels = self.currently_cached_labels.copy()
        self.currently_cached_labels = {}

        self.controller = Controller(max_frame_size=6**24)

    async def _connect_controller(
        self, endpoint: str, username: str, password: str, cacert: str
    ) -> None:
        """Connect to a controller via its endpoint.

        :param str endpoint: the hostname:port endpoint of the controller
            to connect to.
        :param str username: the username for controller-local users
        :param str password: the password for controller-local users
        :param str cacert: the CA certificate of the controller
            (PEM formatted)
        """
        if not self.controller.is_connected():
            await self.controller.connect(
                endpoint=endpoint, username=username, password=password, cacert=cacert
            )

    async def _get_models(self) -> Dict[str, str]:
        """Get a list of all models under a controller.

        :return: str model_uuids: the uuids of all models under the controller
        """
        model_uuids = await self.controller.model_uuids()

        return model_uuids

    async def _get_machines_in_model(self, uuid: str) -> Dict[Any, Any]:
        """Get a list of all machines in the model with their stats.

        :return: status information for all machines in the model
        """
        model = await self.controller.get_model(uuid)
        status = await model.get_status()

        return status["machines"]

    def _create_gauge_label(
        self, hostname: str, model_name: str, machine_type: str
    ) -> Dict[str, str]:
        """Create label dict for gauge.

        :param str hostname: the hostnameof the machine
        :param str model_name: the name of the model the machine is in
        :param str machine_type: the hardware type of the machine
        :return dict labelvalues: the label values in dict format
        """
        return {
            "job": "prometheus-juju-exporter",
            "hostname": hostname,
            "customer": self.config["customer"]["name"].get(str),
            "cloud_name": self.config["customer"]["cloud_name"].get(str),
            "juju_model": model_name,
            "type": machine_type,
        }

    @staticmethod
    def _get_gauge_value(status: str) -> int:
        """Get numerical value for the running status.

        :param str status: the running status of the machine
        :return int status: the integer representation of the running status.
            0 for inactive, 1 for active.
        """
        return int(status == "started")

    def _get_labels_to_remove(self, gauge_name: str) -> None:
        """Get a list of labelvalues for removed machines."""
        stale_labels = {
            k: self.previously_cached_labels[k]
            for k in set(self.previously_cached_labels)
            - set(self.currently_cached_labels)
        }

        for label in stale_labels.values():
            self.data[gauge_name]["labelvalues_remove"].append(list(label[0].values()))

    async def _get_machine_stats(
        self, machines: Dict, model_name: str, gauge_name: str
    ) -> None:
        """Get baremetal or vm machines' stats.

        :param dict machines: status information for all machines in the model
        :param str model_name: the name of the model the machines are in
        :param str gauge_name: the name of the gauge
        """
        for machine in machines.values():
            # check machine type with the prefix of its mac address
            interfaces = machine["network-interfaces"].values()
            machine_type = MachineType.METAL

            for i in interfaces:
                mac_addresses = i["mac-address"]
                if mac_addresses.startswith(tuple(self.prefixes)):
                    machine_type = MachineType.KVM
                    break

            value = self._get_gauge_value(status=machine["agent-status"]["status"])

            labels = self._create_gauge_label(
                hostname=machine["hostname"],
                model_name=model_name,
                machine_type=machine_type.value,
            )

            if labels["hostname"] and labels["hostname"] != "None":
                self.currently_cached_labels[machine["hostname"]] = (
                    labels.copy(),
                    value,
                )

                self.data[gauge_name]["labelvalues_update"].append((labels, value))

            self._get_container_status(
                containers=machine["containers"],
                model_name=model_name,
                gauge_name=gauge_name,
            )

    def _get_container_status(
        self, containers: Dict, model_name: str, gauge_name: str
    ) -> None:
        """Get lxd containers stats.

        :param dict containers: status information for all containers on a machine
        :param str model_name: the name of the model the machines are in
        :param str gauge_name: the name of the gauge
        """
        for container in containers.values():
            value = self._get_gauge_value(container["agent-status"]["status"])

            labels = self._create_gauge_label(
                hostname=container["hostname"],
                model_name=model_name,
                machine_type=MachineType.LXD.value,
            )

            if labels["hostname"] and labels["hostname"] != "None":
                self.currently_cached_labels[container["hostname"]] = (
                    labels.copy(),
                    value,
                )
                self.data[gauge_name]["labelvalues_update"].append((labels, value))

    async def get_stats(self) -> Dict[str, Any]:
        """Get stats from all machines."""
        gauge_name = "juju_machine_state"
        gauge_desc = "Running status of juju machines"
        labels = [
            "job",
            "hostname",
            "customer",
            "cloud_name",
            "juju_model",
            "type",
        ]
        self.refresh_cache(gauge_name=gauge_name, gauge_desc=gauge_desc, labels=labels)

        endpoint = self.config["juju"]["controller_endpoint"].get(str)
        username = self.config["juju"]["username"].get(str)
        password = self.config["juju"]["password"].get(str)
        cacert = self.config["juju"]["controller_cacert"].get(str)

        try:
            await self._connect_controller(
                endpoint=endpoint, username=username, password=password, cacert=cacert
            )
            model_uuids = await self._get_models()
            self.logger.debug("List of models in controller: %s", model_uuids)

            for name, uuid_ in model_uuids.items():
                self.logger.debug("Checking model '%s'...", name)
                machines = await self._get_machines_in_model(uuid=uuid_)
                await self._get_machine_stats(
                    machines=machines, model_name=name, gauge_name=gauge_name
                )

            self._get_labels_to_remove(gauge_name=gauge_name)
        finally:
            await self.controller.disconnect()

        return self.data
